### DISCUSSION TITLE
Future of python-dmidecode (need some maintainers)

### DISCUSSION THREAD

**Discussion by nima** [nima/python-dmidecode#40] [Polls]
Hi All,

I have not had much time to do anything useful on this package, and that's at best caused everyone some frustration I think.  I have started this discussion on how best to move forward from here.  I'm starting this discussion of to see if we can figure something out.

Again, I'm very sorry for my absence, but I'm back now and will do my best to help out.

cc: @lian-bo, @coiby, @vt-alt, @dsommers, @stefanor, @HuO50, @mcepl

Nima

**Comment by lian-bo** [comment]
On Sun, Sep 4, 2022 at 10:40 AM Nima Talebi ***@***.***>
wrote:

> Hi All,
>
> I have not had much time to do anything useful on this package, and that's
> at best caused everyone some frustration I think. I have started this
> discussion on how best to move forward from here. I'm starting this
> discussion of to see if we can figure something out.
>
> Again, I'm very sorry for my absence, but I'm back now and will do my best
> to help out.
>
Good to know, Nima.

Currently, we have finished most of the work about updating SMBIOS-3.3.0,
which is implemented by Zhongze. And we are testing it, once all things are
ready, we
will merge it in the smbion-3.3.0 branch, finally rebase to the master
branch.

BTW: I have helped to review all patches several times offline.

Originally, I would like to tell you after we have finished all the work.
But for now, you have seen a bit of change. :-)

Thanks.
Lianbo

**Reply by nima** [reply]
Ah, thank you Lianbo.  When you have some time, let's have a chat; I want to make sure I do what I can to make things easier for everyone; if there's any changed that you think we should make to how this project is set up, access controls, permissions, groups, and so on.

Feel free to ping me anytime!

**Comment by lian-bo** [comment]
On Sun, Sep 4, 2022 at 11:25 AM Nima Talebi ***@***.***>
wrote:

> Ah, thank you Lianbo. When you have some time, let's have a chat; I want
> to make sure I do what I can to make things easier for everyone; if there's
> any changed that you think we should make to how this project is set up,
> access controls, permissions, groups, and so on.
>
Thank you, Nima.

My initial thought was to do the SMBIOS 3.3.0 update first, and fix some
critical bugs, and then a new release may be in the Mid-November.

And the next work will include(not limit):
[1] improve the test modules
[2] investigate if it's doable to add the json output in python-dmidecode
[3] keep the old xml and convert them between xml and json if possible
[4] closely follow up the upstream dmidecode project
[5] keep an eye on the SMBIOS spec and ensure supporting new features in
new SMBIOS spec
...

We can discuss it after the smbios-3.3.0 branch is ready(update to
SMBIOS-3.3.0), and you can decide if it can be merged in the master branch.
Any thoughts?

Feel free to ping me anytime!
>
I'm using the irc to communicate with people upstream, but I'm not sure if
you  are using it too. If yes, I would suggest creating a python-dmidecode
channel in irc tool,
we can discuss the work via irc channel in the future.

Thanks.
Lianbo

**Comment by dsommers** [comment]
Hi,

I had a period ages ago where I helped co-maintain this project; I'm the one who rewrote lots of the code to get XML support (which was needed in a project I was also involved in, [rteval](https://git.kernel.org/pub/scm/utils/rteval/rteval.git/about/)).  I am now working full time with OpenVPN, so unfortunately my time is too scarce to follow up in this project.

That said, it should be possible to extract JSON out of the main parsing.  If I remember the design, the XML generation happens in two stages.  The first one takes the DMI data and organizes the internal data structure (A Python dict, iirc) based on the [`pymap.xml`](https://github.com/nima/python-dmidecode/blob/master/src/pymap.xml) definition.  That internal data structure is then parsed into the resulting XML.  I would expect this second step to be easily extended outputting JSON based on the internal Python dicts.

Also, iirc, these dicts are accessible via the `dmidecode.QuerySection(...)`.

If you have any questions about these aspects, I can try to help out.  But unfortunately, I don't have much time to spare helping maintaining this project more actively.  At least not before I'm tossed into a project depending on it :slightly_smiling_face:


Also, back in the days, I had several talks with the upstream maintainer of `dmidecode` which this project builds on - to see if dmidecode could be made into a library which could be used to query and retrieve information programmatically.  But back then it was very little interest to do that.

**Comment by lian-bo** [comment]
Hi, David
On Mon, Sep 5, 2022 at 6:54 PM David Sommerseth ***@***.***>
wrote:

> Hi,
>
> I had a period ages ago where I helped co-maintain this project; I'm the
> one who rewrote lots of the code to get XML support (which was needed in a
> project, rteval
> <https://git.kernel.org/pub/scm/utils/rteval/rteval.git/about/>). I am
> now working full time with OpenVPN, so unfortunately my time is too scarce
> to follow up in this project.
>
Thank you for the contribution to python-dmidecode, David, welcome.  :-)

> That said, it should be possible to extract JSON out of the main parsing.
> If I remember the design, the XML generation happens in two stages. The
> first one takes the DMI data and organizes the internal data structure (A
> Python dict, iirc) based on the pymap.xml
> <https://github.com/nima/python-dmidecode/blob/master/src/pymap.xml>
> definition. That internal data structure is then parsed into the resulting
> XML. I would expect this second step to be easily extended outputting JSON
> based on the internal Python dicts.
>
Good news. At least this proves it might be doable in theory.

> Also, iirc, these dicts are accessible via the dmidecode.QuerySection(...)
> .
>
> If you have any questions about these aspects, I can try to help out. But
> unfortunately, I don't have much time to spare helping maintaining this
> project more actively. At least not before I'm tossed into a project
> depending on it 🙂
>
Great. It definitely helps us. :-)

**Comment by lian-bo** [comment]
Hi, Nima

If I remember correctly, two year ago Chenlizhong said that he would like to maintain the python-dmidecode with me together, but later because he had other work and had no much time to work on this project. So I had to find a right people who is interesting in and has enough time to work on the python-dmidecode, eventually, I found HuO50(Zhongze Hu <hzzjoshua@gmail.com>) from my friends.

I have discussed with Zhongze Hu about the python-dmidecode, and he is interesting in the python-dmidecode and has much time to work on this although he was not very familar with the python-dmidecode before this. As you know, Zhongze Hu is working on SMBIOS-3.3.0 update, I only help him to finish the current job, but the main code will be implemented by him.

I would suggest to add Zhongze Hu in the maintainer list after he has finished the current SMBIOS-3.3.0 update. And I will help him to do the maintainace until he could complete all important work on upstream. But anyway, you could make the dicision. :-)

Thanks.
Lianbo

**Reply by HuO50** [reply]
Thanks  Lianbo:
      I am doing something to support python-dmidecode with smbios-3.3. This is an interesting work with much fun. Thanks your support for reviewing and testing patch.  I will continue to develop python-dmidecode .
Thanks 
Zhongze Hu(hzzjoshua@gmail.com)

**Reply by nima** [reply]
Just wanted to say I'm very happy to have you onboard Zhongze Hu, and thank you Lianbo for your contribution and the time you've invested in this project.

**Comment by lian-bo** [comment]
BTW:
Lichen Liu(@lichliu@redhat.com) will do the test for the SMBIOS-3.3.0 update changes. And also thank you all, Zhongze and Lichen.

**Reply by HuO50** [reply]
Thanks Lichen Liu's  work on testing the patch.

Zhongze Hu(hzzjoshua@gmail.com)

**Comment by HuO50** [comment]
Hi Nima & Lianbo:
    Sorry to reply this comments two days later.
    Python-dmidecode is import and interesting, and I am doing some updates on it.  This is my first time to participant open source project, thanks Lian-bo's help on my reviewing and testing, he gave me lots of useful tips. Now I will do some work on updating, but i don't have enough machines to test my work, anyone who know how to solve it, please tell me.
Thanks
Zhongze Hu(hzzjoshua@gmail.com)

**Reply by dsommers** [reply]
Inside the [unit-tests/private](https://github.com/nima/python-dmidecode/tree/master/unit-tests/private) directory, there are a lot of various "dump files".  They contain data which can be loaded using `dmidecode.set_dev()`.

```python
import dmidecode

dmidecode.set_dev('private/IBM-x3950-M2.1.dmidump')
dmidecode.QueryTypeId(0x001)
```

This directory can be extended with more DMI dumps.  To capture DMI data from a new machine:
```python
import dmidecide

dmidecode.set_dev('dump.dmi')
dmidecode.dump()
```

Give the file a model name (instead of `dump.dmi`), similar to the files already in the `private/` directory, and commit that file.  Then it will automatically be run with the unit-test.

Collecting DMI data from newer machines would be a good thing, especially those models adding more DMI data not found in these old files.

**Reply by lian-bo** [reply]
> Hi Nima & Lianbo: Sorry to reply this comments two days later. Python-dmidecode is import and interesting, and I am doing some updates on it. This is my first time to participant open source project, thanks Lian-bo's help on my reviewing and testing, he gave me lots of useful tips. Now I will do some work on updating, but i don't have enough machines to test my work, anyone who know how to solve it, please tell me. Thanks Zhongze Hu([hzzjoshua@gmail.com](mailto:hzzjoshua@gmail.com))

No worries, Zhongze. Lichen Liu(@[lichliu@redhat.com](mailto:lichliu@redhat.com)) can help to do the test in the future, I asked him. Thanks.

**Reply by HuO50** [reply]
@dsommers Thanks dsmmers, I am using the method you suggest to test the newest patch and fix some bugs. thanks you.
Also thanks @lian-bo and @lichliu for supporting.

**Comment by lian-bo** [comment]
Hi, @nima 
Sorry to disturb you again.  :-)  Could  you help sync the branch smbios-3.3.0 code to the master branch? Seems that I have no permission to rebase from the smbions-3.3.0 branch to the master branch, I did it according to the following steps and got a failure:

[1] [python-dmidecode]$ git rebase smbios-3.3.0 master

[2] [python-dmidecode]$ git push
Username for 'https://github.com': lian-bo
Password for 'https://lian-bo@github.com': 
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: At least 1 approving review is required by reviewers with write access.
To https://github.com/nima/python-dmidecode.git
 ! [remote rejected]           master -> master (protected branch hook declined)
error: failed to push some refs to 'https://github.com/nima/python-dmidecode.git'

Or [3]submit a pull request as below:
https://github.com/nima/python-dmidecode/pull/44

Thanks
Lianbo

**Comment by nima** [comment]
Hi Lianbo,

Not disturbing me at all, of course I will do that.

It's late now, but i'll do it in the morning first thing. I'll also check
the permissions, and fix your access.

Nima

**Reply by lian-bo** [reply]
Thank you for the help, Nima. It's not urgent.

Currently,  Zhongze Hu is fixing several known issues, and  Lichen Liu will help with the test, although they are not critical issues.

I guess that some known issues can be solved befor the new release this time. :-)

**Comment by nima** [comment]
Hi Lianbo,

Okasy, this seems to be due to how master has been set up (see
screenshot).  I think you have access to modify this setting too. I haven't
change it yet, because I wanted to talk to you about it first.  Seems like
a good idea to leave it this way, and simply get the approval from a
collaborator, what do you think?

Nima
[image: Screen Shot 2022-10-24 at 9.20.37 AM.png]

**Reply by lian-bo** [reply]
Thank you for sharing your idea, Nima.

I agree that the current way does have many advantages, the most important point is that it guarantees the code quality and prevents arbitrary modification because the submitted code(big changes) will be reviewed by another reviewer.

However, this requires at least two collaborators, otherwise the code can not be merged(just like https://github.com/nima/python-dmidecode/pull/44) unless we modifiy the github setting.

In fact, the collaborators can reach the following consensus: at lease one ACK is required, but the permission for merging is still open. This still has the current advantages. In the future, if the other collaborators become inactive, maybe I can still move on. :-) Does this look more flexible or efficient? Any thoughts?

BTW: I can not download the AM.png(your screen shot), maybe my brower is problematic.  :-)

Thanks.
Lianbo

**Reply by lian-bo** [reply]
In addition, I would recommend adding Zhongze Hu(hzzjoshua@gmail.com) and Lichen Liu(@lichliu@redhat.com) in collaborator/maintainer list. Maybe Zhongze will try to maintain the code parts and release, etc, Lichen will help with the test. If needed, I will assist them with the above.

But anyway, you can make a decision. :-)

Thank you in advance, Nima.

Lianbo

**Reply by nima** [reply]
Taking a look at this now Lianbo; will comment back here soon.

**Reply by nima** [reply]
I've sent out an invite to both:
- Zhongze Hu <hzzjoshua@gmail.com>
- Lichen Liu <lichliu@redhat.com>

If we have enough collaborators, having 2 collaborators ACK might be feasible, but I'm not sure if we have enough, even with Lichen and Zhongze.  I think I'll reduce the required number of collaborators to 1 for now, and if the community grows, then we can increase it again to 2.  What do you think?

**Reply by nima** [reply]
Hmmm.... I just checked, the `master` branch requires 1 approval right now, where are you seeing 2 required?

**Reply by lian-bo** [reply]
> I've sent out an invite to both:
> 
> * Zhongze Hu [hzzjoshua@gmail.com](mailto:hzzjoshua@gmail.com)
> * Lichen Liu [lichliu@redhat.com](mailto:lichliu@redhat.com)
> 
> If we have enough collaborators, having 2 collaborators ACK might be feasible, but I'm not sure if we have enough, even with Lichen and Zhongze. I think I'll reduce the required number of collaborators to 1 for now, and if the community grows, then we can increase it again to 2. What do you think?

This sounds good to me. Thank you so much, Nima.

**Reply by lian-bo** [reply]
> Hmmm.... I just checked, the `master` branch requires 1 approval right now, where are you seeing 2 required?

You are right, Currently the master branch requires one approval.

My question is that I have no permission to rebase from the smbios-3.3.0 branch to the master branch, in this case, it requires that another collaborator helps to review and give a approval, otherwise it will be blocked.

But for now, it doesn't have any problems because I have asked Zhongze to help resubmit the PR, I can do it and merge it into the master branch.

Thanks.
Lianbo

**Reply by HuO50** [reply]
Hello , all.
      Thanks @nima  and @lian-bo.  I ack this collaborators invite, and I will continue working on it.

**Reply by HuO50** [reply]
hello, nima 
      My gmail not received the invite mail. At first I think it may  caused by network delay, but after 7 days, I checked in the gmail trash and filters, nor have the invite mail. Can you sent the invite mail again? @nima

**Reply by nima** [reply]
Okay done!  Sorry for the late response Huzz =)

Let me know if anything else comes up!

**Reply by HuO50** [reply]
Tks, @nima , I had received mail this time.

**Reply by nima** [reply]
Thanks for letting me know!
